### PR TITLE
K8s: standalone: use Grafana's logger to stream all logs

### PR DESCRIPTION
**Pull request by charandas**
**What is this feature?**

The standalone config continues to be different from embedded APIServer in Grafana. While we work on aligning these, the immediate need is to allow for specifying logging options / interfaces as developers are used to.

This changeset pulls in the ExtraOptions from within the grafana-apiserver package and uses it to control a few things.

1. Logging
2. Dev mode: with the addition of this, we can make the file write for loopback config to be conditional.

**Why do we need this feature?**

[Add a description of the problem the feature is trying to solve.]

**Who is this feature for?**

[Add information on what kind of user the feature is for.]

**Which issue(s) does this PR fix?**:

Fixes #

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: pkg/cmd/grafana/apiserver/cmd.go [hunk 1]
```diff
@@ -67,16 +78,55 @@ func newCommandStartExampleAPIServer(o *APIServerOptions, stopCh <-chan struct{}
...
-	o.RecommendedOptions = options.NewRecommendedOptions(
-		defaultEtcdPathPrefix,
-		grafanaapiserver.Codecs.LegacyCodec(), // the codec is passed to etcd and not used
-	)
 	o.RecommendedOptions.AddFlags(cmd.Flags())
 
 	return cmd
 }
 
+func setupLogging(logLevel int) error {
```
Comment (marefr) [inline pkg/cmd/grafana/apiserver/cmd.go:89]: With my other comment would suggest moving this function to the pkg/infra/log package to keep the ini file/logic at the same place. We can name it something like `SetupConsoleLogger`. No blocker for this PR though

File: pkg/cmd/grafana/apiserver/cmd.go [hunk 2]
```diff
@@ -30,6 +36,11 @@ func newCommandStartExampleAPIServer(o *APIServerOptions, stopCh <-chan struct{}
 			devAcknowledgementNotice,
 		Example: "grafana apiserver --runtime-config=example.grafana.app/v0alpha1=true",
 		RunE: func(c *cobra.Command, args []string) error {
+			// Setup logging as the very first thing after we have read the desired verbosity level
+			if err := setupLogging(o.ExtraOptions.Verbosity); err != nil {
```
Comment (marefr) [inline pkg/cmd/grafana/apiserver/cmd.go:40]: With my other comments, I think we can setup the Grafana logger even earlier, in the https://github.com/grafana/grafana/blob/1208888bb618774228aefe65884e06316a06e714/pkg/cmd/grafana/apiserver/cmd.go#L80

**Review by marefr — APPROVED** [review]
Nice LGTM. Added some suggestions that we can do in a follow up or here, I'll leave that up to you

File: pkg/cmd/grafana/apiserver/cmd.go [hunk 3]
```diff
@@ -85,3 +135,48 @@ func RunCLI() int {
 
 	return cli.Run(cmd)
 }
+
+var _ logr.LogSink = (*logAdapter)(nil)
```
Comment (toddtreece) [inline pkg/cmd/grafana/apiserver/cmd.go:139]: does this need to be copied? not sure if this is the issue, but the log adapter could be made public

Comment (marefr) [reply pkg/cmd/grafana/apiserver/cmd.go:139]: What about https://github.com/grafana/grafana/pull/84530#discussion_r1526166311?

File: pkg/cmd/grafana/apiserver/cmd.go (hunk 1 above)
Comment (marefr) [reply pkg/cmd/grafana/apiserver/cmd.go]: I moved it

File: pkg/cmd/grafana/apiserver/cmd.go (hunk 2 above)
Comment (marefr) [reply pkg/cmd/grafana/apiserver/cmd.go]: Kept the setup in the command for now

File: pkg/cmd/grafana/apiserver/cmd.go (hunk 3 above)
Comment (marefr) [reply pkg/cmd/grafana/apiserver/cmd.go]: I re-used extra options log wrapper by calling `ExtraOptions.Apply(...)`

File: pkg/cmd/grafana/apiserver/cmd.go [hunk 4]
```diff
@@ -67,16 +78,55 @@ func newCommandStartExampleAPIServer(o *APIServerOptions, stopCh <-chan struct{}
...
+	if err != nil {
+		fmt.Println(err)
+		return nil
+	}
+
+	logger := logr.New(newLogAdapter(logLevel))
+	klog.SetLoggerWithOptions(logger, klog.ContextualLogger(true))
+	if _, err := logs.GlogSetter(strconv.Itoa(logLevel)); err != nil {
+		logger.Error(err, "failed to set log level")
+	}
```
[the remark this answers is not in the document]
Comment (marefr) [reply pkg/cmd/grafana/apiserver/cmd.go]: Done

File: pkg/cmd/grafana/apiserver/cmd.go [hunk 5]
```diff
@@ -30,6 +29,14 @@ func newCommandStartExampleAPIServer(o *APIServerOptions, stopCh <-chan struct{}
 			devAcknowledgementNotice,
 		Example: "grafana apiserver --runtime-config=example.grafana.app/v0alpha1=true",
 		RunE: func(c *cobra.Command, args []string) error {
+			if err := log.SetupConsoleLogger("debug"); err != nil {
```
Comment (marefr) [inline pkg/cmd/grafana/apiserver/cmd.go:32]: Debug hardcoded for now. Question if we should map verbosity level to grafana log level, i.e. if verbosity
- 0 = info
- 1 = warning
- 2 = error
- 3 = error
- >= 4 = debug

Based on https://github.com/kubernetes/klog/blob/007e661a007dacdf017ad9c1b53e9f96d32d4b7d/internal/severity/severity.go#L30-L36

**Review by marefr — APPROVED** [review]
LGTM, but I'm biased since I just pushed some changes :)

**Review by toddtreece — APPROVED** [review]

File: pkg/cmd/grafana/apiserver/cmd.go (hunk 5 above)
Comment (toddtreece) [reply pkg/cmd/grafana/apiserver/cmd.go:32]: > We can probably figure this out in a separate PR

yeah, i think this is going to take some additional research. my understanding of the levels is limited, but it seems like they are closer to increasing the debug output

**Pull request merged**